algebra/ring.lean

```diff
@@ -15,6 +15,85 @@ section
   (left_distrib n 1 1).trans (by simp)
 end
 
+section units
+  variables [semiring α]
+  variable (α)
+
+  structure invertible :=
+    (val : α)
+    (inv : α)
+    (property : val * inv = 1 ∧ inv * val = 1)
+
+  variables {a b c : invertible α}
+  variable {α}
+
+  def ext (HAB : a.val = b.val) : a = b :=
+  begin
+    cases a,
+    cases b,
+    dsimp at HAB,
+    congr,
+    exact HAB,
+    exact calc
+        inv = inv * (val_1 * inv_1) : by rw [property_1.1, mul_one]
+        ... = (inv * val_1) * inv_1 : by rw [mul_assoc]
+        ... = (inv * val) * inv_1   : by rw [HAB]
+        ... = inv_1                 : by rw [property.2, one_mul]
+  end
+
+  lemma mul_four {a b c d : α} :
+  (a * b) * (c * d) = a * (b * c) * d := by simp
+
+  def mul (a b : invertible α) : invertible α :=
+  {
+    val := a.val * b.val,
+    inv := b.inv * a.inv,
+    property := ⟨by rw [mul_four,b.property.1,mul_one,a.property.1],
+                by rw [mul_four,a.property.2,mul_one,b.property.2]⟩
+  }
+
+  def one : invertible α :=
+  {
+    val := 1,
+    inv := 1,
+    property := ⟨mul_one 1, one_mul 1⟩
+  }
+
+  def inv (a : invertible α) : invertible α :=
+  {
+    val := a.inv,
+    inv := a.val,
+    property := ⟨a.property.2, a.property.1⟩
+  }
+
+  instance : has_mul (invertible α) := ⟨mul⟩
+  instance : has_one (invertible α) := ⟨one⟩
+  instance : has_inv (invertible α) := ⟨inv⟩
+
+  variables (a b c)
+
+  @[simp] lemma mul_simp : (mul a b).val = a.val * b.val := rfl
+  @[simp] lemma one_simp : (one : invertible α).val = 1 := rfl
+  @[simp] lemma inv_simp : (inv a).val = a.inv := rfl
+  @[simp] lemma mul_simp' : (a*b).val = a.val * b.val := rfl
+  @[simp] lemma one_simp' : (1 : invertible α).val = 1 := rfl
+  @[simp] lemma inv_simp' : (a⁻¹).val = a.inv := rfl
+  @[simp] lemma mul_inv_simp' : (a*a⁻¹).val = 1 := by simp [a.property.1]
+  @[simp] lemma inv_mul_simp' : (a⁻¹*a).val = 1 := by simp [a.property.2]
```

Comment:
You shouldn't have any theorems called *_simp. Try to write them so that they are useful when called explicitly as well as by simp. My suggestions: `mul_val`, `one_val`, `inv_val`. I don't think you need `mul_inv_simp'`, `mul_inv` (the global one) will cover this. You can avoid `mul_simp` and `one_simp` by inlining the definition of `mul` and `one` into the instances, and then referring exclusively to `1` and `*` instead.

Reply:
@kckennylau also why do you have a theorem about `(mul a b).val` and one for `(a * b).val`? I would recommend to _not_ define `mul` but only `has_mul`.